src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1040 @@
+       (genlo, genhi) = genRange rng
+       b = fromIntegral genhi - fromIntegral genlo + 1 :: Integer
+
+       -- Probabilities of the most likely and least likely result
+       -- will differ at most by a factor of (1 +- 1/q). Assuming the RandomGen
+       -- is uniform, of course
+
+       -- On average, log q / log b more pseudo-random values will be generated
+       -- than the minimum
+       q = 1000 :: Integer
+       k = h - l + 1
+       magtgt = k * q
+
+       -- generate pseudo-random values until we exceed the target magnitude
+       f mag v g | mag >= magtgt = (v, g)
+                 | otherwise = v' `seq`f (mag*b) v' g' where
+                        (x,g') = next g
+                        v' = v * b + (fromIntegral x - fromIntegral genlo)
+
+-- | Generate an integral in the range @[l, h]@ if @l <= h@ and @[h, l]@
+-- otherwise.
+uniformIntegralM :: (Bits a, Integral a, StatefulGen g m) => (a, a) -> g -> m a
+uniformIntegralM (l, h) gen = case l `compare` h of
+  LT -> do
+    let limit = h - l
+    let limitAsWord64 :: Word64 = fromIntegral limit
+    bounded <-
+      if fromIntegral limitAsWord64 == limit
+        -- Optimisation: if 'limit' fits into 'Word64', generate a bounded
+        -- 'Word64' and then convert to 'Integer'
+        then fromIntegral <$> unsignedBitmaskWithRejectionM uniformWord64 limitAsWord64 gen
+        else boundedExclusiveIntegralM (limit + 1) gen
+    return $ l + bounded
+  GT -> uniformIntegralM (h, l) gen
+  EQ -> pure l
+{-# INLINEABLE uniformIntegralM #-}
+
+-- | Generate an integral in the range @[0, s)@ using a variant of Lemire's
+-- multiplication method.
+--
+-- Daniel Lemire. 2019. Fast Random Integer Generation in an Interval. In ACM
+-- Transactions on Modeling and Computer Simulation
+-- https://doi.org/10.1145/3230636
+--
+-- PRECONDITION (unchecked): s > 0
+boundedExclusiveIntegralM :: forall a g m . (Bits a, Integral a, StatefulGen g m) => a -> g -> m a
+boundedExclusiveIntegralM (s :: a) gen = go
+  where
+    n = integralWordSize s
+    -- We renamed 'L' from the paper to 'k' here because 'L' is not a valid
+    -- variable name in Haskell and 'l' is already used in the algorithm.
+    k = wordSizeInBits * n
+    twoToK = (1::a) `shiftL` k
+    modTwoToKMask = twoToK - 1
+
+    t = (twoToK - s) `mod` s
+    go :: (Bits a, Integral a, StatefulGen g m) => m a
+    go = do
+      x <- uniformIntegralWords n gen
+      let m = x * s
+      -- m .&. modTwoToKMask == m `mod` twoToK
+      let l = m .&. modTwoToKMask
+      if l < t
+        then go
+        -- m `shiftR` k == m `quot` twoToK
+        else return $ m `shiftR` k
+{-# INLINE boundedExclusiveIntegralM #-}
+
+-- | @integralWordSize i@ returns that least @w@ such that
+-- @i <= WORD_SIZE_IN_BITS^w@.
+integralWordSize :: (Bits a, Num a) => a -> Int
+integralWordSize = go 0
+  where
+    go !acc i
+      | i == 0 = acc
+      | otherwise = go (acc + 1) (i `shiftR` wordSizeInBits)
+{-# INLINE integralWordSize #-}
```

Comment:
One can a) detect the size of finite integer types in compile time, b) detect the size of `Integer`/`Natural` in O(1) time, if `integer-gmp` is available.

```haskell
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE CPP          #-}
{-# LANGUAGE MagicHash    #-}

import Data.Bits
import GHC.Exts

#if defined(MIN_VERSION_integer_gmp)
import GHC.Integer.GMP.Internals
import GHC.Natural
#else
import GHC.Integer
import GHC.Natural
#endif

wordSize :: Int
wordSize = finiteBitSize (0 :: Word)

-- | @integralWordSize i@ returns that least @w@ such that
-- @i <= WORD_SIZE_IN_BITS^w@.
integralWordSize :: (Bits a, Integral a) => a -> Int
integralWordSize n = case bitSizeMaybe n of
  Just s -> (s + wordSize - 1) `quot` wordSize
#if defined(MIN_VERSION_integer_gmp)
  Nothing
    | isSigned n -> case toInteger n of {- n appeared to be an Integer -}
      S#{} -> 1
      Jp# bn -> I# (sizeofBigNat# bn)
      Jn# bn -> I# (sizeofBigNat# bn)
    | otherwise -> case fromIntegral n of {- n appeared to be a Natural -}
      NatS#{} -> 1
      NatJ# bn -> I# (sizeofBigNat# bn)
#else
  Nothing -> go 0 n
  where
    go !acc i
      | i == 0 = acc
      | otherwise = go (acc + 1) (i `shiftR` wordSize)
#endif
{-# INLINE integralWordSize #-}
```

Reply:
This is a very nice idea, but I would like to suggest to postpone it to after this release, there is already plenty of optimizations here I don't want to get stuck on implementing more. Let's leave a little room for improvement ;)

@Bodigrim At the same time if you or anyone else really wants to get it done now, I will not stand in a way.

Reply:
Agreed. I might get time to implement it before `random-1.2`, but this one (and the one below) are not blockers.